### DISCUSSION TITLE
Adding PF_ROUTE failed

### DISCUSSION THREAD

**Discussion by MadhukarMKB** [strongswan/strongswan#2079] [General]
Hi, 
I am using sun, moon setup for host-host connectivity

I am trying on QNX platform, with pf_key and pf_route enabled and kernel-netlink disabled 
My swanctl.conf (moon):
connections {
   host-host {
      local_addrs  = 10.220.84.110
      remote_addrs = 10.220.84.120

      local {
         auth = pubkey
         certs = moonCert.pem
         id = moon.strongswan.org
      }
      remote {
         auth = pubkey
         id = sun.strongswan.org
      }
      children {
         host-host {
            updown = /home/mkumarasw/strongswan/EPREFIX/libexec/ipsec/_updown iptables
			policies = yes
            rekey_time = 5400
            rekey_bytes = 500000000
            rekey_packets = 1000000
			esp_proposals = aes128ctr-sha1
         }
      }
      version = 2
      mobike = no
      reauth_time = 10800
	  proposals = aes128-sha1-x25519,aes128-sha-x25519,aes128-sha512-x25519,aes128-sha2_384-x25519,
   }
}

SUN; swanctl.conf:
connections {
   host-host {
      local_addrs  = 10.220.84.120
      remote_addrs = 10.220.84.110

      local {
         auth = pubkey
         certs = sunCert.pem
         id = sun.strongswan.org
      }
      remote {
         auth = pubkey
         id = moon.strongswan.org
      }
      children {
         host-host {
            updown = /home/mkumarasw/strongswan/EPREFIX/libexec/ipsec/_updown iptables
			policies = yes
            rekey_time = 5400
            rekey_bytes = 500000000
            rekey_packets = 1000000
			esp_proposals = aes128ctr-sha1	
         }
      }
      version = 2
      mobike = no
      reauth_time = 10800
      proposals =  aes128-sha1-x25519,aes128-sha-x25519,aes128-sha512-x25519,aes128-sha2_384-x25519
   }
}

The IKE is established successfully.
Jan 24 06:27:44 13[IKE] authentication of 'sun.strongswan.org' (myself) with RSA_EMSA_PKCS1_SHA2_256 successful
Jan 24 06:27:44 13[IKE] sending end entity cert "C=CH, O=strongswan, CN=sun.strongswan.org"
Jan 24 06:27:44 13[IKE] IKE_SA host-host[1] established between 10.220.84.120[sun.strongswan.org]...10.220.84.110[moon.strongswan.org]
Jan 24 06:27:44 13[IKE] IKE_SA host-host[1] state change: CONNECTING => ESTABLISHED

But looking some error in CHILD_SA

Jan 24 06:27:44 13[CHD] integrity responder key => 20 bytes @ 0x4822c2b0
Jan 24 06:27:44 13[CHD]    0: 45 4E 2B DB D2 BB 7A 11 9A DD 73 B2 64 B6 44 16  EN+...z...s.d.D.
Jan 24 06:27:44 13[CHD]   16: FA C9 C6 E1                                      ....
Jan 24 06:27:44 13[CHD] adding inbound ESP SA
Jan 24 06:27:44 13[CHD]   SPI 0xdcee9af8, src 10.220.84.110 dst 10.220.84.120
Jan 24 06:27:44 13[ESP] adding SAD entry with SPI dcee9af8 and reqid {1}
Jan 24 06:27:44 13[ESP]   using encryption algorithm AES_CTR with key size 160
Jan 24 06:27:44 13[ESP]   using integrity algorithm HMAC_SHA1_96 with key size 160
Jan 24 06:27:44 13[CHD] adding outbound ESP SA
Jan 24 06:27:44 01[JOB] next event in 22s 815ms, waiting
Jan 24 06:27:44 13[CHD]   SPI 0x5c478d78, src 10.220.84.120 dst 10.220.84.110
Jan 24 06:27:44 13[ESP] adding SAD entry with SPI 5c478d78 and reqid {1}
Jan 24 06:27:44 13[ESP]   using encryption algorithm AES_CTR with key size 160
Jan 24 06:27:44 13[ESP]   using integrity algorithm HMAC_SHA1_96 with key size 160
Jan 24 06:27:44 01[JOB] next event in 22s 815ms, waiting
Jan 24 06:27:44 13[ESP] adding policy 10.220.84.110/32 === 10.220.84.120/32 in
Jan 24 06:27:44 13[ESP] adding policy 10.220.84.120/32 === 10.220.84.110/32 out
Jan 24 06:27:44 13[KNL] getting a local address in traffic selector 10.220.84.120/32
Jan 24 06:27:44 13[KNL] using host 10.220.84.120
Jan 24 06:27:44 07[JOB] events on fds: 15[r]
Jan 24 06:27:44 07[JOB] observing fds: 18[r]
Jan 24 06:27:44 07[JOB] watcher is observing 1 fds
Jan 24 06:27:44 13[KNL] using 10.220.84.110 as nexthop to reach 10.220.84.110
Jan 24 06:27:44 13[KNL] installing route: 10.220.84.110/32 src 10.220.84.120 dev tun0
Jan 24 06:27:44 07[JOB] watcher got notification, rebuilding
Jan 24 06:27:44 07[JOB] observing fds: 15[r] 18[r]
Jan 24 06:27:44 07[JOB] watcher is observing 2 fds
Jan 24 06:27:44 13[KNL] adding PF_ROUTE route failed: Network is unreachable
Jan 24 06:27:44 07[JOB] events on fds: 15[r]
Jan 24 06:27:44 13[KNL] installing route failed: 10.220.84.110/32 src 10.220.84.120 dev tun0
Jan 24 06:27:44 07[JOB] observing fds: 18[r]
Jan 24 06:27:44 13[IKE] unable to install IPsec policies (SPD) in kernel
Jan 24 06:27:44 13[IKE] failed to establish CHILD_SA, keeping IKE_SA
Jan 24 06:27:44 07[JOB] watcher is observing 1 fds
Jan 24 06:27:44 13[CHD] CHILD_SA host-host{1} state change: INSTALLING => DESTROYING

**Comment by tobiasbrunner** [comment]
Why is there a `tun0` device?

**Reply by MadhukarMKB** [reply]
It is trying to create the tunnel
NE120:1-40*:/tmp # route show
Routing tables

Internet:
Destination        Gateway            Flags
0.0.0.0/32         link#3             U
10.0.0.0/8         10.220.84.1        UG
10.220.84.0/23     link#10            U
10.220.84.1        c4:7e:e0:7e:19:1d  UHL
20.0.0.0/24        link#13            U
100.100.84.120/32  link#12            U

**Reply by MadhukarMKB** [reply]
tun0 is initialized from:
get_tun_name() in the below call
451     INIT(route,
452         .if_name = router->get_tun_name(router, is_virtual ? src_ip : NULL),
453         .src_ip = src_ip,
454         .dst_net = chunk_clone(policy->dst.net->get_address(policy->dst.net)),
455         .prefixlen = policy->dst.mask,
456     );

**Reply by tobiasbrunner** [reply]
Only if you'd use the [_kernel-libipsec_ plugin](https://docs.strongswan.org/docs/5.9/plugins/kernel-libipsec.html), but you haven't mentioned that (and if you want to use the IPsec implementation in the OS kernel that's not the way to go anyway). The _kernel-pfroute_ plugin can also create TUN devices, but only if virtual IPs are used, which your config does not indicate, it's rather a simple host-to-host connection that should not require that.

**Reply by MadhukarMKB** [reply]
Yes, I am using kernel-libipsec, kernel-pfkey, kernel-pfroute, I did not tried with virtual IP as it is host-host it i snot required.
Is there any config miss ?
Please guide on this to resolve
My configure looks like:
./configure --host=ppc64 --prefix=$HOME/strongswan/PREFIX --exec-prefix=$HOME/strongswan/EPREFIX --enable-shared --disable-systemd --disable-static --enable-monolithic --with-urandom-device=/dev/urandom --disable-gmp  --with-linux-headers --enable-kernel-libipsec --enable-openssl --enable-kernel-pfroute --enable-kernel-pfkey --disable-kernel-netlink --enable-ikev2 --disable-attr --disable-resolve --disable-stroke --disable-xauth-generic --enable-aes --enable-des --disable-rc2 --disable-md5 --enable-constraints --enable-pubkey --disable-pgp --disable-dnskey --disable-sshkey --disable-fips-prf --disable-xcbc --disable-cmac --enable-hmac --enable-socket-default --enable-sha1 --enable-sha2 --enable-openssl --enable-kernel-macsec --enable-event-notify --enable-socket-dynamic  --enable-curve25519 --enable-pkcs12 --disable-pkcs7 --enable-pkcs1 --enable-random  --enable-pem --enable-pkcs8 --enable-kdf --enable-x509 --enable-revocation --enable-drbg --enable-counters

**Reply by tobiasbrunner** [reply]
Yes, don't use that plugin. Why would you? Did you even read the docs?

**Reply by MadhukarMKB** [reply]
With libipsec disabled I observed the peer charon got crashed

**Reply by tobiasbrunner** [reply]
So debug it.

**Reply by MadhukarMKB** [reply]
I tried with disabling the kernel-libipsec, 
But I am getting some error while pfkey_send it seems:
Jan 25 05:37:22 09[KNL]   using encryption algorithm AES_CTR with key size 160
Jan 25 05:37:22 09[KNL]   using integrity algorithm HMAC_SHA1_96 with key size 160
**Jan 25 05:37:22 09[KNL] unable to add SAD entry with SPI c4bd9768: Invalid argument (22)**
Jan 25 05:37:22 09[CHD] adding outbound ESP SA
Jan 25 05:37:22 09[CHD]   SPI 0xcca52d3c, src 10.220.84.120 dst 10.220.84.110
Jan 25 05:37:22 09[KNL] adding SAD entry with SPI cca52d3c and reqid {1}
Jan 25 05:37:22 09[KNL]   using encryption algorithm AES_CTR with key size 160
Jan 25 05:37:22 09[KNL]   using integrity algorithm HMAC_SHA1_96 with key size 160
**Jan 25 05:37:22 09[KNL] unable to add SAD entry with SPI cca52d3c: Invalid argument (22)**
Jan 25 05:37:22 09[IKE] unable to install inbound and outbound IPsec SA (SAD) in kernel

**Reply by tobiasbrunner** [reply]
Yes, the kernel is not happy with something in the SADB_ADD message. Could be the algorithms (e.g. AES-CTR is rather exotic), but there are a lot of additional information passed with the message (or maybe it's also something missing). So you'd have to check in the kernel to see where and why it complains.

**Reply by MadhukarMKB** [reply]
Hi, 

The crash is not happening now.
But I am seeing something like 

But I am getting some error while pfkey_send it seems:
Jan 25 05:37:22 09[KNL] using encryption algorithm AES_CTR with key size 160
Jan 25 05:37:22 09[KNL] using integrity algorithm HMAC_SHA1_96 with key size 160
**Jan 25 05:37:22 09[KNL] unable to add SAD entry with SPI c4bd9768: Invalid argument (22)**
Jan 25 05:37:22 09[CHD] adding outbound ESP SA
Jan 25 05:37:22 09[CHD] SPI 0xcca52d3c, src 10.220.84.120 dst 10.220.84.110
Jan 25 05:37:22 09[KNL] adding SAD entry with SPI cca52d3c and reqid {1}
Jan 25 05:37:22 09[KNL] using encryption algorithm AES_CTR with key size 160
Jan 25 05:37:22 09[KNL] using integrity algorithm HMAC_SHA1_96 with key size 160
**Jan 25 05:37:22 09[KNL] unable to add SAD entry with SPI cca52d3c: Invalid argument (22)**
Jan 25 05:37:22 09[IKE] unable to install inbound and outbound IPsec SA (SAD) in kernel

And I see pfkey socket working as prescribed, and the kernel replying with invalid argument (22)

 tried with aes192ctr-sha1 aes128ctr-sha1, and aes128ctr-md5, for all I am observing the same reply
 
 BTW I am using strongswan on QNX platform

**Reply by MadhukarMKB** [reply]
Hi, 
Is there any way that I can check the command format (actual command) sending to kernel and actual reply coming from kernel ?
